storage: proactively renew expiration-based leases

a-robinson opened the pull request:
This mechanism won't scale incredibly well to proactively renewing all
expiration-based leases since it'd require a goroutine per lease. It
should be fine if we only auto-renew r1's lease, though.

Fixes #24753

Release note: None

----------------

@bdarnell, do you have any suggestions for testing this? I've verified it manually, but it's hard to properly isolate this code in a realistic way. We could ratchet the lease durations way down and wait until the lease's expiration bumps, but that creates a timing-based test that is either flaky if we use a really short lease duration or slow if we use a long one.

a-robinson committed:
storage: proactively renew r1's lease

Run a store-level background worker that proactively renews the lease
for r1. It could easily be extended to also renew other expiration-based
leases if we so desire in the future.

Fixes #24753

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25322)
<!-- Reviewable:end -->

bdarnell commented:
Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---

*[pkg/storage/replica.go, line 1544 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25322#-LBqs9S907F5_JF09qxR:-LBqs9S907F5_JF09qxS:b49p5x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d0998fabc9c1186f3be13d98823fa6924bbdd85/pkg/storage/replica.go#L1544)):*
> ```Go
> 			}
> 
> 			// redirectOnOrAcquireLease doesn't wait until the lease request
> ```

I don't like the level of detailed knowledge of the lease request machinery here. Could we do something simpler like send a Get request to Replica.Send and let all the lease stuff happen under the covers?

---

*[pkg/storage/replica_proposal.go, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25322#-LBqs_rG5O87bStkZ_pt:-LBqs_rG5O87bStkZ_pu:b2xd84e) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d0998fabc9c1186f3be13d98823fa6924bbdd85/pkg/storage/replica_proposal.go#L274)):*
> ```Go
> 		// different replicas as maybeGossipFirstRange is called on each (e.g.
> 		// #24753).
> 		r.store.Stopper().RunWorker(r.AnnotateCtx(context.Background()), func(ctx context.Context) {
> ```

I'm slightly worried that we could have a lot of these goroutines pile up if they're not exiting for some reason. I was imagining a single store-level goroutine that would periodically check to see whether we have a replica of r1 and if so try to do something to refresh the lease. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25322)*
<!-- Sent from Reviewable.io -->

a-robinson committed:
storage: Proactively renew all expiration-based leases

This builds off the previous commit which proactively renewed
only the first range's lease.

Release note: None

bdarnell commented:
LGTM

a-robinson commented:
bors r+

a-robinson commented:
bors r+
